File: Samples/Geolocation/js/geolocationjs/js/scenario2_GetPosition.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved
//// This code is licensed under the MIT License (MIT).

(function () {
    "use strict";

    var getGeopositionPromise;
    var desiredAccuracyInMeters = 0;

    var disposed;
    var page = WinJS.UI.Pages.define("/html/scenario2_GetPosition.html", {
        ready: function (element, options) {
            disposed = false;
            document.getElementById("getGeopositionButton").addEventListener("click", getGeoposition, false);
            document.getElementById("cancelGetGeopositionButton").addEventListener("click", cancelGetGeoposition, false);
            document.getElementById("getGeopositionButton").disabled = false;
            document.getElementById("cancelGetGeopositionButton").disabled = true;
            document.getElementById("desiredAccuracyInMeters").addEventListener("keyup", getDesiredAccuracy);
        },
        dispose: function () {
            if (!disposed) {
                disposed = true;
                if (document.getElementById("getGeopositionButton").disabled) {
                    if (getGeopositionPromise) {
                        getGeopositionPromise.operation.cancel();
                    }
                }
            }
        }
    });

    function getGeoposition() {
        document.getElementById("getGeopositionButton").disabled = true;
        document.getElementById("cancelGetGeopositionButton").disabled = false;

        Windows.Devices.Geolocation.Geolocator.requestAccessAsync().done(
            function (accessStatus) {
                switch (accessStatus) {
                    case Windows.Devices.Geolocation.GeolocationAccessStatus.allowed:

                        // If DesiredAccuracy or DesiredAccuracyInMeters are not set (or value is 0), DesiredAccuracy default is used.
                        var geolocator = new Windows.Devices.Geolocation.Geolocator();
                        geolocator.desiredAccuracyInMeters = desiredAccuracyInMeters;

                        getGeopositionPromise = geolocator.getGeopositionAsync();
                        getGeopositionPromise.done(
                            function (pos) {
                                if (!disposed) {
                                    WinJS.log && WinJS.log("Location updated", "sample", "status");

                                    updateLocationData(pos);
                                    document.getElementById("getGeopositionButton").disabled = false;
                                    document.getElementById("cancelGetGeopositionButton").disabled = true;
                                }
                            },
                            function (err) {
                                if (!disposed) {
                                    WinJS.log && WinJS.log(err.message, "sample", "error");

                                    updateLocationData(null);
                                    document.getElementById("getGeopositionButton").disabled = false;
                                    document.getElementById("cancelGetGeopositionButton").disabled = true;
                                }
                            }
                        );

                        WinJS.log && WinJS.log("Waiting for update...", "sample", "status");
                        document.getElementById("getGeopositionButton").disabled = true;
                        document.getElementById("cancelGetGeopositionButton").disabled = false;
                        break;

                    case Windows.Devices.Geolocation.GeolocationAccessStatus.denied:
                        WinJS.log && WinJS.log("Access to location is denied.", "sample", "error");
                        break;

                    case Windows.Devices.Geolocation.GeolocationAccessStatus.unspecified:
                        WinJS.log && WinJS.log("Unspecified error!", "sample", "error");
                        break;
                }
            },
            function (err) {
                WinJS.log && WinJS.log(err, "sample", "error");
            });
    }

    function cancelGetGeoposition() {
        if (getGeopositionPromise) {
            getGeopositionPromise.operation.cancel();
        }
        document.getElementById("getGeopositionButton").disabled = false;
        document.getElementById("cancelGetGeopositionButton").disabled = true;
    }

    function getDesiredAccuracy(evt) {
        var e = document.getElementById("desiredAccuracyInMeters");
        var stringValue = e.value;
        var value = parseInt(stringValue);
        var nullAllowed = true;
        var elementName = "Desired accuracy";

        if (!value) {
            var msg;
            // value is either empty or alphabetic
            if (0 !== stringValue.length) {
                WinJS.log && WinJS.log("Desired accuracy must be a number.", "sample", "error");
            }
        } else {
            // update value
            desiredAccuracyInMeters = value;

            // clear out status message
            WinJS.log && WinJS.log("", "sample", "status");
        }
    }

    function updateLocationData(pos) {
        if (pos == null) {
            document.getElementById("latitude").innerHTML = "No data";
            document.getElementById("longitude").innerHTML = "No data";
            document.getElementById("accuracy").innerHTML = "No data";
            document.getElementById("source").innerHTML = "No data";

            // hide labels and satellite data
            document.getElementById("positionLabel").style.opacity = 0;
            document.getElementById("position").style.opacity = 0;
            document.getElementById("horizontalLabel").style.opacity = 0;
            document.getElementById("horizontal").style.opacity = 0;
            document.getElementById("verticalLabel").style.opacity = 0;
            document.getElementById("vertical").style.opacity = 0;
        } else {
            var coord = pos.coordinate;
            document.getElementById("latitude").innerHTML = coord.point.position.latitude;
            document.getElementById("longitude").innerHTML = coord.point.position.longitude;
            document.getElementById("accuracy").innerHTML = coord.accuracy;
            document.getElementById("source").innerHTML = toPositionSource(coord.positionSource);

            if (coord.positionSource === Windows.Devices.Geolocation.PositionSource.satellite) {
                // show labels and satellite data
                document.getElementById("positionLabel").style.opacity = 1;
                document.getElementById("position").style.opacity = 1;
                document.getElementById("horizontalLabel").style.opacity = 1;
                document.getElementById("horizontal").style.opacity = 1;
                document.getElementById("verticalLabel").style.opacity = 1;
                document.getElementById("vertical").style.opacity = 1;
                document.getElementById("position").innerHTML = coord.satelliteData.positionDilutionOfPrecision;
                document.getElementById("horizontal").innerHTML = coord.satelliteData.horizontalDilutionOfPrecision;
                document.getElementById("vertical").innerHTML = coord.satelliteData.verticalDilutionOfPrecision;
            } else {
                // hide labels and satellite data
                document.getElementById("positionLabel").style.opacity = 0;
                document.getElementById("position").style.opacity = 0;
                document.getElementById("horizontalLabel").style.opacity = 0;
                document.getElementById("horizontal").style.opacity = 0;
                document.getElementById("verticalLabel").style.opacity = 0;
                document.getElementById("vertical").style.opacity = 0;
            }
        }
    }

    function toPositionSource(val) {
        switch (val) {
            case Windows.Devices.Geolocation.PositionSource.satellite:
                return "Satellite";
            case Windows.Devices.Geolocation.PositionSource.wiFi:
                return "Wifi";
            case Windows.Devices.Geolocation.PositionSource.cellular:
                return "cellular";
            case Windows.Devices.Geolocation.PositionSource.ipAddress:
                return "IPAddress";
            case Windows.Devices.Geolocation.PositionSource.unknown:
                return "Unknown";
        }
    }
})();

File: Samples/Geolocation/cppwinrt/Scenario2_GetPosition.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario2_GetPosition.h"
#include "Scenario2_GetPosition.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace winrt::Windows::Devices::Geolocation;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario2_GetPosition::Scenario2_GetPosition()
    {
        InitializeComponent();
    }

    void Scenario2_GetPosition::OnNavigatedFrom(NavigationEventArgs const&)
    {
        if (m_pendingOperation)
        {
            m_pendingOperation.Cancel();
            m_pendingOperation = nullptr;
        }
    }

    fire_and_forget Scenario2_GetPosition::GetGeolocationButtonClicked(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        GetGeolocationButton().IsEnabled(false);
        CancelGetGeolocationButton().IsEnabled(true);
        LocationDisabledMessage().Visibility(Visibility::Collapsed);

        try
        {
            // Request permission to access location
            auto accessStatus = co_await Geolocator::RequestAccessAsync();

            switch (accessStatus)
            {
            case GeolocationAccessStatus::Allowed:
            {
                m_rootPage.NotifyUser(L"Waiting for update...", NotifyType::StatusMessage);

                // If DesiredAccuracy or DesiredAccuracyInMeters are not set (or value is 0), DesiredAccuracy.Default is used.
                Geolocator geolocator;
                uint32_t desiredAccuracyInMetersValue;
                if (TryParseUInt(DesiredAccuracyInMeters().Text().c_str(), desiredAccuracyInMetersValue))
                {
                    geolocator.DesiredAccuracyInMeters(desiredAccuracyInMetersValue);
                }

                // Carry out the operation
                m_pendingOperation = geolocator.GetGeopositionAsync();
                Geoposition pos = co_await m_pendingOperation;

                UpdateLocationData(pos);
                m_rootPage.NotifyUser(L"Location updated.", NotifyType::StatusMessage);
            }
                break;

            case GeolocationAccessStatus::Denied:
                m_rootPage.NotifyUser(L"Access to location is denied.", NotifyType::ErrorMessage);
                LocationDisabledMessage().Visibility(Visibility::Visible);
                UpdateLocationData(nullptr);
                break;

            case GeolocationAccessStatus::Unspecified:
                m_rootPage.NotifyUser(L"Unspecified error.", NotifyType::ErrorMessage);
                UpdateLocationData(nullptr);
                break;
            }
        }
        catch (hresult_canceled const&)
        {
            m_rootPage.NotifyUser(L"Canceled.", NotifyType::StatusMessage);
        }
        catch (...)
        {
            m_rootPage.NotifyUser(to_message(), NotifyType::ErrorMessage);
        }

        m_pendingOperation = nullptr;
        GetGeolocationButton().IsEnabled(true);
        CancelGetGeolocationButton().IsEnabled(false);
    }

    void Scenario2_GetPosition::CancelGetGeolocationButtonClicked(IInspectable const&, RoutedEventArgs const&)
    {
        if (m_pendingOperation)
        {
            m_pendingOperation.Cancel();
            m_pendingOperation = nullptr;
        }
        GetGeolocationButton().IsEnabled(true);
        CancelGetGeolocationButton().IsEnabled(false);
    }

    // Updates the user interface with the Geoposition data provided
    void Scenario2_GetPosition::UpdateLocationData(Geoposition const& position)
    {
        if (position == nullptr)
        {
            ScenarioOutput_Latitude().Text(L"No data");
            ScenarioOutput_Longitude().Text(L"No data");
            ScenarioOutput_Accuracy().Text(L"No data");
            ScenarioOutput_Source().Text(L"No data");
            ShowSatelliteData(false);
        }
        else
        {
            ScenarioOutput_Latitude().Text(to_hstring(position.Coordinate().Point().Position().Latitude));
            ScenarioOutput_Longitude().Text(to_hstring(position.Coordinate().Point().Position().Longitude));
            ScenarioOutput_Accuracy().Text(to_hstring(position.Coordinate().Accuracy()));
            ScenarioOutput_Source().Text(to_hstring(position.Coordinate().PositionSource()));

            if (position.Coordinate().PositionSource() == PositionSource::Satellite)
            {
                // Show labels and satellite data when available
                GeocoordinateSatelliteData satelliteData = position.Coordinate().SatelliteData();
                ScenarioOutput_PosPrecision().Text(to_hstring(satelliteData.PositionDilutionOfPrecision()));
                ScenarioOutput_HorzPrecision().Text(to_hstring(satelliteData.HorizontalDilutionOfPrecision()));
                ScenarioOutput_VertPrecision().Text(to_hstring(satelliteData.VerticalDilutionOfPrecision()));;
                ShowSatelliteData(true);
            }
            else
            {
                // Hide labels and satellite data
                ShowSatelliteData(false);
            }
        }
    }

    void Scenario2_GetPosition::ShowSatelliteData(bool isVisible)
    {
        auto visibility = isVisible ? Visibility::Visible : Visibility::Collapsed;
        ScenarioOutput_PosPrecision().Visibility(visibility);
        ScenarioOutput_HorzPrecision().Visibility(visibility);
        ScenarioOutput_VertPrecision().Visibility(visibility);
    }

}
